ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -901,101 +900,95 @@ deriving instance SOP.HasDatatypeInfo (ImmDB.TraceEvent e)
-    genAppendToCurrentChain :: Gen TestHash
-    genAppendToCurrentChain = do
-        forkNo <- genForkNo
-        return $ TestHash $ case Block.pointHash $ Model.tipPoint dbModel of
-          GenesisHash            ->         forkNo NE.:| []
-          BlockHash (TestHash h) -> NE.cons forkNo h
-
-    -- A hash that fits onto a block in the ChainDB. The block could be at the
-    -- tip of the chain and the hash might already be present in the ChainDB.
-    genFitsOnSomewhere :: Gen TestHash
-    genFitsOnSomewhere = do
-        hashInDB <- unTestHash <$> elements hashesInDB
-        forkNo <- genForkNo
-        return $ TestHash $ NE.cons forkNo hashInDB
-
-    -- A hash that doesn't fit onto a block in the ChainDB, but it creates a
+    -- A block that already exists in the ChainDB
+    genAlreadyInChain :: Gen TestBlock
+    genAlreadyInChain = elements $ Map.elems blocksInChainDB
+
+    -- A block that fits onto the current chain
+    genAppendToCurrentChain :: Gen TestBlock
+    genAppendToCurrentChain = case Model.tipBlock dbModel of
+      Nothing -> genFirstBlock
+      Just b  -> genFitsOn b
+
+    -- A block that fits onto some block @b@ in the ChainDB. The block @b@
+    -- could be at the tip of the chain and the generated block might already
+    -- be present in the ChainDB.
+    genFitsOnSomewhere :: Gen TestBlock
+    genFitsOnSomewhere = case Model.tipBlock dbModel of
+      Nothing -> genFirstBlock
+      Just _  -> genAlreadyInChain >>= genFitsOn
+
+    -- A block that doesn't fit onto a block in the ChainDB, but it creates a
     -- gap of a couple of blocks between genesis or an existing block in the
-    -- ChainDB.
-    genGap :: Gen TestHash
+    -- ChainDB. We generate it by generating a few intermediary blocks first,
+    -- which we don't add. But the chance exists that we will generate them
+    -- again later on.
+    genGap :: Gen TestBlock
     genGap = do
-        hash <- frequency
-          [ (if empty then 0 else 3,
-             NE.toList . unTestHash <$> elements hashesInDB)
-            -- Genesis
-          , (1, return [])
-          ]
-        gapSize <- choose (1, 2) -- TODO relate it to @k@?
-        -- ForkNos for the gap
-        forkNos <- replicateM gapSize genForkNo
-        return $ TestHash $ NE.fromList $ foldr (:) hash forkNos
+        gapSize <- choose (1, 3)
+        start   <- genFitsOnSomewhere
+        go gapSize start
+      where
+        go :: Int -> TestBlock -> Gen TestBlock
+        go 0 b = return b
+        go n b = genFitsOn b >>= go (n - 1)
+
+    -- Generate a block or EBB fitting on genesis
+    genFirstBlock :: Gen TestBlock
+    genFirstBlock = frequency
+      [ (1, firstBlock <$> chooseSlot 0 2 <*> genBody)
+      , (4, firstEBB <$> genBody)
+      ]
+
+    -- Helper that generates a block that fits onto the given block.
+    genFitsOn :: TestBlock -> Gen TestBlock
+    genFitsOn b = frequency
+        [ (4, do
```

Comment:
Comment: If we generate too many EBBs, a consequence will be that the density of the chain will be low.

Reply:
Let's add a comment saying that while the test block _type_ is fairly general (and could indeed perhaps even be used in the consensus tests), this _generator_ is specific to the chain DB test.